pkg/machines/libvirt-dbus.es6

```diff
@@ -0,0 +1,861 @@
+
+        let xmlDesc = getDiskXML(diskFileName, target);
+
+        return () => {
+            clientLibvirt[connectionName].call(vmId, 'org.libvirt.Domain', 'AttachDevice', [xmlDesc, flags], TIMEOUT)
+                    .fail(ex => console.error("VM ATTACH DISK action for disk %s failed: %s", diskFileName, JSON.stringify(ex)));
+        };
+    },
+
+    CHANGE_NETWORK_STATE({
+        name,
+        id: objPath,
+        networkMac,
+        state,
+        connectionName
+    }) {
+        return dispatch => {
+            clientLibvirt[connectionName].call(objPath, 'org.libvirt.Domain', 'GetXMLDesc', [0], TIMEOUT)
+                    .done(domXml => {
+                        let updatedXml = updateNetworkIfaceState(domXml[0], networkMac, state);
+                        if (!updatedXml) {
+                            dispatch(vmActionFailed({
+                                name,
+                                connectionName,
+                                message: _("VM CHANGE_NETWORK_STATE action failed: updated device XML couldn't not be generated"),
+                            }));
+                        } else {
+                            clientLibvirt[connectionName].call(objPath, 'org.libvirt.Domain', 'UpdateDevice', [updatedXml, 0], TIMEOUT)
+                                    .done(() => {
+                                        dispatch(getVm({connectionName, id:objPath}));
+                                    })
+                                    .fail(exception => dispatch(vmActionFailed({
+                                        name,
+                                        connectionName,
+                                        message: _("VM CHANGE_NETWORK_STATE action failed"),
+                                        detail: {exception}
+                                    })));
+                        }
+                    });
+        };
+    },
+
+    CREATE_AND_ATTACH_VOLUME({
+        connectionName,
+        poolName,
+        volumeName,
+        size,
+        format,
+        target,
+        vmId,
+        permanent,
+        hotplug
+    }) {
+        let volXmlDesc = getVolumeXML(volumeName, size, format, target);
+
+        return dispatch => {
+            clientLibvirt[connectionName].call('/org/libvirt/QEMU', 'org.libvirt.Connect', 'StoragePoolLookupByName', [poolName], TIMEOUT)
+                    .done((storagePoolPath) => {
+                        clientLibvirt[connectionName].call(storagePoolPath[0], 'org.libvirt.StoragePool', 'StorageVolCreateXML', [volXmlDesc, 0], TIMEOUT)
+                                .done((storageVolumePath) => {
+                                    clientLibvirt[connectionName].call(storageVolumePath[0], "org.freedesktop.DBus.Properties", "GetAll", ["org.libvirt.StorageVol"], TIMEOUT)
+                                            .done(function(props) {
+                                                if (!$.isEmptyObject(props[0]))
```

Comment:
Please replace with non-jQuery alternative: `Object.getOwnPropertyNames(props[0]).length > 0`

Reply:
Done

Reply:
So in this case we only need volume path, I would suggest using Get method instead of GetAll.  In addition wherever we use GetAll this check is not sufficient enough because the returned props can contain only subset of all available properties so it may happen that the one we need is not included and the condition would still be true.  GetAll is best effort dbus call and it doesn't have to return all properties, it returns what it can.

Reply:
Done